### PR TITLE
Diff source tree after interactive build

### PR DESCRIPTION
**Pull request by Keno**
Implements https://github.com/JuliaPackaging/BinaryBuilder.jl/issues/125.

Example PR: https://github.com/JuliaPackaging/Yggdrasil/pull/519

### CODE REVIEW

File: src/wizard/interactive_build.jl [hunk 1]
```diff
@@ -136,44 +136,101 @@ end
...
+        cmd = `/bin/bash -c $cmd`
+    end
+
+    run_interactive(ur, cmd, stdin=state.ins, stdout=state.outs, stderr=state.outs)
+
+    had_patches = false
+    if srcdir_overlay
+        # Check if there were any modifications made to the srcdir
+        if length(readdir(joinpath(prefix, "srcdir"))) > 0
+            # Note that we're mounting to /workspace/srcdir/upper so comparing
```
Comment (staticfloat) [inline src/wizard/interactive_build.jl]: Is this comment right?  It looks like you're mounting to `/workspace/srcdir`, not `/workspace/srcdir/upper`.

File: src/wizard/deploy.jl [hunk 2]
```diff
@@ -107,7 +112,7 @@ end
 Write out a WizardState to a `build_tarballs.jl` in an `Yggdrasil` clone, then
 open a pull request against `Yggdrasil`.
 """
-function yggdrasil_deploy(state::WizardState)
+function yggdrasil_deploy(state::WizardState, open_it::Bool = true)
```
Comment (staticfloat) [inline src/wizard/deploy.jl]: Instead of `open_it` let's call it `open_pr`.

File: src/wizard/deploy.jl [hunk 3]
```diff
@@ -199,15 +222,22 @@ function step7(state::WizardState)
...
             "Write to a local file",
             "Print to stdout",
         ])
     )
     println(state.outs)
 
     if deploy_select == 1
-        yggdrasil_deploy(state)
+        yggdrasil_select = request(terminal,
+            "Would you like to actuall open the pull request or just prepare it?",
```
Comment (staticfloat) [inline src/wizard/deploy.jl]:
```suggestion
            "Would you like to actually open the pull request or just prepare it?",
```

File: src/Prefix.jl [hunk 4]
```diff
@@ -338,13 +348,23 @@ function setup_workspace(build_path::AbstractString, sources::Vector{<:SetupSour
     wrapperdir = joinpath(workspace, "compiler_wrappers")
     mkdir.((srcdir, destdir, metadir))
 
+    if srcdir_overlay
+        mkdir(joinpath(build_path, nonce, "srcdir_work"))
+        srcdir = joinpath(build_path, nonce, "srcdir_lower")
+        mkdir(srcdir)
+    end
```
Comment (staticfloat) [inline src/Prefix.jl]: I would prefer that this is always `srcdir`, not `srcdir_lower`; reduces complexity and makes it so I don't have to think about what's going on.  We can bindmount it to `${WORKSPACE}/srcdir_lower` just before we mount the overlay to get this to work.

File: src/Sources.jl [hunk 5]
```diff
@@ -107,6 +107,10 @@ function SetupSource(url::String, path::String, hash::String, target::String)
     end
 end
 
+struct PatchSource
+    name::String
+    patch::String
+end
```
Comment (staticfloat) [inline src/Sources.jl]: I'm a little worried someone will use this in their `build_tarballs.jl`; we probably want to give this a different name to make sure it's an internal-only thing.

Comment (giordano) [reply src/Sources.jl]: Note that it isn't an `AbstractSource` (nor it is exported), so someone using it as source in a builder won't go too far

File: src/Prefix.jl [hunk 6]
```diff
@@ -313,6 +313,16 @@ function setup(source::SetupSource{DirectorySource}, targetdir, verbose)
     end
 end
 
+function setup(source::PatchSource, targetdir, verbose)
+    # Need to strip the trailing separator also here
```
Comment (giordano) [inline src/Prefix.jl]: Copied by mistake?
```suggestion
```

File: src/wizard/deploy.jl [hunk 7]
```diff
@@ -1,15 +1,20 @@
...
         elseif any(endswith(x[1], ext) for ext in archive_extensions)
             "ArchiveSource($(url_string), $(repr(x[2].hash)))"
         else
-            "FileSource($(url_string), $(repr(x[2].hash))),"
+            "FileSource($(url_string), $(repr(x[2].hash)))"
         end
-    end,"\n    "))
+    end
+    if !isempty(state.patches)
+        push!(sources_strings, "\"./bundled\"")
```
Comment (giordano) [inline src/wizard/deploy.jl]:
```suggestion
        push!(sources_strings, "DirectorySource(\"./bundled\")")
```

**Pull request merged**